### ISSUE TITLE
Can't find swiflint

### ISSUE THREAD

**Issue by Javarome** [bealex/SwiftLintAppCode#55]
I had the plugin working on my project for a while, then it stopped working with that message:
```
Can't find swiftlint utility here:
/opt/homebrew/bin/swiftlint
Please check the path in settings.
```
`/opt/homebrew/bin/swiftlint` is the result of `which swiftlint`. Running `swiftlint` or `/opt/homebrew/bin/swiftlint` on the command line works.

My `.swiflint.yml` is at the root of my project.

Setting `/opt/homebrew/bin` in Tools > Swiftlint > Swiftlint path doesn't work either.

- Plugin version: 1.13.3
- AppCode version: 2021.1.3
- Swiflint version: 0.43.1

**Comment by bealex** [comment]
@Javarome Please check with new version. Hope I got this.

**Comment by Javarome** [comment]
Thanks @bealex. Got the 1.13.4 and, well, yes and no: no more errors indeed, swiftlint seems to be found now... but no swiftlint warnings are displayed in the editor (only JetBrains inspections). Again, such warning are properly listed using command line swiftlint in the same directory.

**Comment by bealex** [comment]
@Javarome Is SwiftLint is turned on in inspections?

**Comment by Javarome** [comment]
Yes it was disabled then I enabled it again. 
<img width="637" alt="Capture d’écran 2021-07-20 à 01 20 20" src="https://user-images.githubusercontent.com/2269113/126239576-c5ab2240-7c22-49bf-9b93-0cbf1dfa646e.png">
But still no warnings.
I tried to relaunch AppCode as well, with no more luck.

**Comment by Javarome** [comment]
@bealex Actually I can see only _one_ swiftlint warning, about trailing_whitespace.

So obviously it is running, but I wonder if it is find/using the proper `.swiflint.yml`. Mine is located at the root of the project.

**Comment by bealex** [comment]
Yup, it is something with config locations, seeing same in some projects. Will get to it soon.

**Comment by bealex** [comment]
Just to be sure I'm looking in the right direction. How many projects do you have opened at the same time?

**Comment by Javarome** [comment]
Only one. 

For what it's worth, the config file is `~/AppCodeProjects/myproject/.swiftlint.yml`.

But of course there are many others in libs caches (like `~/Library/Caches/JetBrains/AppCode2021.1/DerivatedData/MyProject/-xxxx/SourcePackages/checkouts/SomeLib/.swiftlint.yml`) and externs (like `Extern/Sparkle/.swiftlint.yml` for instance)

**Comment by bealex** [comment]
Can you please try this version? This way it will be faster then go through JB approval process. https://www.dropbox.com/s/ulwpmf2ew9fbxff/SwiftLint.zip?dl=0

**Comment by Javarome** [comment]
Installed this 1.13.5 zip from disk but nothing better, sorry.

**Comment by bealex** [comment]
Thanks! OK, thinking...

**Comment by Javarome** [comment]
Maybe some additional configuration file path in the plugin settings could help? If left blank, default config files lookup would apply, otherwise it would be passed as `swiflint --config` arguments.

**Comment by bealex** [comment]
Right now I'm not passing `--config` parameter, it breaks some scenarios. Basically what I'm doing is:
 - getting source root directory (for example project root)
 - getting file that we are linting now and directory that it is contained in
Somewhere in between these directories there can be SwiftLint configuration file (it must be called `.swiftlint.yml`, I can't deal with other config file names)

I try to find this file (starting from source root), and run `swiftlint` in the directory where config file is found. If no file is found, `swiftlint` is run from the source root directory.

This should work for project files, for editing of frameworks (that can be added as sources and contain their own source roots). And it will ensure all `swiftlint` logic about using several/nested `.swiftlint.yml` files works.

It would help a lot if you describe where (from the project root) your config files are located, and where sources are located.

**Comment by Javarome** [comment]
Okay here is an excerpt from my project structure:
- `~/project/myproject` Running `swiftlint` from here outputs 6 violations, 0 serious in 460 files
  - `.swiftlint.yml`
  - `Myproject/` Running `swiftlint` from here outputs 1861 violations, 650 serious in 412 files
    - `.idea/`
    - `Classes/` Running `swiftlint` from here outputs 1651 violations, 493 serious in 401 files
      - `Components/` Running `swiftlint` from here outputs 788 violations, 303 serious in 152 files
        - `ComponentA/` Running `swiftlint` from here outputs 42 violations, 11 serious in 14 files 
          - `Subdir/` Running `swiftlint` from here outputs 42 violations, 11 serious in 14 files 
            - `SomeClass.swift` Running `swiftlint SomeClass.swift` from `Subdir/` outputs  9 violations, 0 serious in 1 file
  - `Myproject.xcodeproj`
  - `Myproject.xcworkspace`
  - `Module1/`
  - `Module1Tests/`
  - `MyprojectTests/`
  - `MyprojectUITests/`
  - `build`
  etc

There are aussi libs that have their own swiftlint config, like:
- `~/Library/Caches/JetBrains/AppCode2021.1/DerivedData/Myproject-xxx/SourcePackages/checkouts`
  - `SwiftCheck` 
  - `swifter`
    - `.swiftlint.yml`
    - `DangerFile` contains `swiftlint.config_file = '.swiftlint.yml'`

**Comment by Javarome** [comment]
Hi @bealex ,

Just to let you know: I stumbled on a file that does display more Swiflint warnings in the IDE. Those are 4 warnings:
<img width="1137" alt="swiftlint ide warnings" src="https://user-images.githubusercontent.com/2269113/127988377-08f1f97e-41c9-4906-9ef7-c9280140e70c.png">

However, running swiftlint cli on the same files yields **9 results**: the 4 control_statement warnings + 4 line length violation warnings (>120) + 1 line length error (>200)

So again it looks like some swiftlint config is active, but not the same as used by cli.

**Comment by bealex** [comment]
Thanks a lot!

I'm slacking, have to keep up with my day job :)

Should return to the investigation in couple of days.

**Comment by markeissler** [comment]
Any updates for this issue? I'm trying to use 0.13.4 with AppCode 2021.3.2, `swiftlint` is 0.45. I have a `.swiftlint.yml` file at the root of my SPM project. Running swiftlint from the command line works. 

If I run `Inspect code...` swiftlint doesn't appear to run at all. The only output in the Problems view are from the existing AppCode Swift inspection rules. I also don't get any swiftlint output in the editor whatsoever (that is, nothing is flagged with red squiggly lines).

Swiftlint is enabled in Inspections as per the setup instructions.

At this point, this plugin does nothing.

**Comment by markeissler** [comment]
I have since updated the plugin to 0.13.5 (the version posted in this thread). I also updated `swiftlint` to `0.46.1`.

Then, I created a new inspections profile that disables everything except Swiftlint. I ran `Inspect code...` on a file. From the terminal I ran...

```bash
ps auxw | grep -i swiftlint
```

This revealed a single `swiftlint` process had started. If I run `switftlint` from the terminal, and point it at my config file, it complets almost instantly. I decided to kill the process in the terminal. Another one started. I killed that one and output was immediately shown in the `Problem` view in AppCode. The editor view itself was not updated.

Hrmm.

I edited the file in the editor using a known trigger for the colon rule, and then also threw in a very long line as a comment.

I went back to the terminal and once again could see a `swiftlint` process had begun. ~I killed it. Another one appeared.~ I killed it and immediately the editor view was updated to highlight errors from the `swiftlint` output.

Seems to me that for some reason, your plugin is launching `swiftlint` but then it's just hanging. 

This is what the process looks like from the terminal:

```bash
mark             91796   0.0  0.1 35020376  27220   ??  S     1:40PM   0:00.32 /usr/local/Cellar/swiftlint/0.46.2/bin/swiftlint lint --no-cache --reporter csv --path /Users/mark/dev/Company/Project/Sources/App/Controllers/UsersController.swift
```

To clarify, that is running on a single file selected via the `Inspect code...` menu option.

~It's interesting that `swiftlint` is restarted if I kill it, but just restarted once. Still, I have to kill it twice to get output in AppCode.~

EDIT: I tested this a bit more. I think it is just one instance of `swiftlint` that is started unless  the inspector is triggered a second time while waiting for the first instance to complete. If I make a single change in the editor that should trigger `swiftlint` to output a linting violation, then as soon as I kill that single instance the editor updates with the appropriate violation flagged.

**Comment by markeissler** [comment]
Something that just occurred to me, if the plugin ends up calling `swiftlint` like this:

```bash
/usr/local/Cellar/swiftlint/0.46.2/bin/swiftlint lint --no-cache --reporter csv --path /Users/mark/dev/Company/Project/Sources/App/Controllers/UsersController.swift
```

Then this plugin could never work for linting while typing since changes haven't necessarily been saved to disk yet. What needs to happen is that the content of the editor is piped into stdin of the `swiftlint` command which would have the `--use-stdin` flag set.

Another thing I noticed is that while this plugin doesn't seem to work when trying to read a file on the `--path` option, it DOES work if you pipe the file (via cat) into the command itself.

I have created a Gist for this as a demo. Works for the following instances...

- linting via `Inspect code...`
- auto linting after file save

https://gist.github.com/markeissler/bcc99d4e7b3e59be5e57c67ebbf77251

**Comment by bealex** [comment]
@markeissler Hi. Thanks for the investigation and for the response and for the code. Appreciate it a lot.

Looking into hanging now. This is something I did not consider for some strange reason.

There are fundamental problems with stdin option though.
 - It does not account for file filters that swiftlint is configured with (included/excluded). 
 - It does not anything about configuration files hierarchy (and it can be quite complex).
 - And, as far as I remember, it also gives swiftlint less information about other files in the module (that can be used by SourceKit to make better decisions about code).

I tried that solution. It was good at the start, but unfortunately it proved itself to be quite brittle and hard to maintain.

I can easily be missing something though. Please tell me if I do.

**Comment by bealex** [comment]
@markeissler Can you please try to install this new version from the disk? Maybe I've found what was causing this behavior (still can't reproduce it), a test will be helpful.

https://github.com/bealex/SwiftLintAppCode/blob/bugfix/swiftlint-hanging/deployed/SwiftLint.zip

**Comment by markeissler** [comment]
All of your points are valid. I'm just trying to find some sort of workaround.

I just tried your latest build from the post above. Same result. `swiftlint` is launched soon after opening a file. But it never exits. As soon as I kill it, results appear in the editor window.

It must be reading the file contents since it does display linting errors in the editor. But with that said, it doesn't appear to complete even though it returns results after killing the process. In a particular file, I get 439 linter warnings using the stdin approach, using your updated binary I get around 200 warnings. One of the ones I checked was the `trailing_whitespace` rule. I guess more of those errors when using the stdin approach.

I think that's telling us that it's possible that swiftlint itself is encountering problems? I'm running the latest version of that as mentioned above (0.46.1).

**Comment by markeissler** [comment]
I guess I could try and learn enough kotlin to try and debug this.

**Comment by markeissler** [comment]
FWIW I can't seem to get `0.13.4` working even with stdin. I have to use either of the builds posted in this thread so far.

**Comment by bealex** [comment]
@markeissler Thanks for the test. Looks like swiftlint does not close the streams (either output or error) for some strange reason. Or output there is non-consistent.

Is it possible to get that file? Maybe it is a part of some open source project? I'd like to reproduce this a lot.

After your test I think I know what happens. Let me try to implement a fast hack to find out am I right.

**Comment by markeissler** [comment]
>Is it possible to get that file? Maybe it is a part of some open source project? I'd like to reproduce this a lot.

Unfortunately, it's closed source, proprietary.

**Comment by bealex** [comment]
@markeissler I've updated the plugin here: https://github.com/bealex/SwiftLintAppCode/blob/bugfix/swiftlint-hanging/deployed/SwiftLint.zip can you please test it once again?

**Comment by markeissler** [comment]
I think with this one you set a timeout of 1 second, is this correct? On smaller files it appears to work but on larger files (I'm looking at a file that is over 2800 lines) I think it might be exiting too early.

**Comment by bealex** [comment]
> I think with this one you set a timeout of 1 second, is this correct?

Yes it is. It's just a test.

> On smaller files it appears to work but on larger files (I'm looking at a file that is over 2800 lines) I think it might be exiting too early.

Thanks! Now when I know what happens, it'll be easy to fix. :-)

**Comment by bealex** [comment]
@markeissler I've updated the plugin, sent the version to the Marketplace. Will be available after JB review. You can try it directly from the repository if needed: https://github.com/bealex/SwiftLintAppCode/raw/main/deployed/SwiftLint.zip

@Javarome can you please test it too? It is possible that you've encountered same error and new build will fix that.

**Comment by Javarome** [comment]
Thanks @markeissler and @bealex for investigating again on that issue! Unfortunately I installed the zip plugin from disk but can't see any swiftlint warnings in my IDE sources (whereas some are raised when using the command line), even after AppCode restart.

For the record: Swiflint 0.46.2 in /opt/homebrew/bin/swiftlint, Swiftlint plugin 1.13.5 (configured with same swiftlint path), swiftlint inspection enabled.

**Comment by bealex** [comment]
@Javarome Heh. OK, will need to test more then.

**Comment by markeissler** [comment]
@Javarome Try creating a new `Inspection Profile` that only enables `SwiftLint`. This is just to isolate this particular inspection from the others and to reduce the amount of analysis that is performed by AppCode.

Then, perhaps test this on a smaller file. Wait for analysis to perform and end. You can see this in the top-right of the editor. It will say something like `Analyzing...` and if you hover over it you'll see progress. Syntax analysis can take a while to complete. Also, Indexing must have finished in order for analysis to begin.

~The other point to keep in mind is that linting will only begin once a file has saved since `swiftlint` is passed a path to the file.~ I don't think this is correct. I'm seeing linting performed even without saving the file, so I'm not sure how this works.

Another thing I've just done is to disable all of the AppCode-provided Swift inspections. The only inspection I've left on (for Swift) is SwiftLint. I'm not sure if there is any point in having both SwiftLint inspections and the built-in inspections for Swift enabled (I would think that the `.swiftlint.yml` configuration should cover everything. Doing this improves performance.

@bealex One thing that would help is if you could change the version number between test builds. e.g. `1.5.0-test1`, `1.5.0-test2`. I've been switching between zip files for testing and the version number installed is always the same which makes it difficult to know which build I have installed.

I also cloned this project and opened it in IntelliJ thinking I could help debug this. But the project doesn't seem to follow the structure that is documented (at least these days). It doesn't look like a gradle or devkit project. There's no make file either. Any pointers would be helpful. :)

![Screen Shot 2022-02-07 at 8 32 21 AM](https://user-images.githubusercontent.com/2679087/152830740-e0e17319-9675-4fbd-a199-dc8e21e8b77f.png)

**Comment by bealex** [comment]
@markeissler I'll add development section to readme. Project is quite old, maybe its time to update it as well :)

**Comment by bealex** [comment]
@markeissler I've added a small dev-related section to the readme. You can try. Wanted to port it to Gradle plugin, but unfortunately it does not support AppCode plugins. So DevKit it is. For now :-)

**Comment by markeissler** [comment]
Excellent! I'll try to take a look later today.

>Wanted to port it to Gradle plugin, but unfortunately it does not support AppCode plugins. 

How ironic: I thought they were trying to get away from DevKit.

**Comment by bealex** [comment]
@Javarome One more question. Are your swift sources in "Project Sources and Headers" directory? You can see all project sources in Project tool window (View → Tool Windows → Project), and by selecting in its header "Project Source Files".

**Comment by Javarome** [comment]
Yes @bealex they are: Project Source Files lists 2 root directories :
- my /Library/Caches/JetBrains/AppCode2021.3/DerivedData/etc... /CoreDataGenerated (not sure this is correct)
- my project root directory (and sub-directories with swift files)

**Comment by bealex** [comment]
Ok. Back to thinking.

**Comment by Javarome** [comment]
@markeissler I did set an inspection profile with only `Swiftlint (available for Code|Inspect Code)`. Then inspection goes green (whereas cli swiflint still has warnings).

**Comment by markeissler** [comment]
@bealex I've opened up a PR branched off of `main` that makes a small change to reading of the inputStream returned by `swiftlint`.

https://github.com/bealex/SwiftLintAppCode/pull/59

This appears to work on my system in testing. I'll get more real use in today.

@Javarome You might want to test the build artifact in that PR (under the `deployed` directory). After install it will have a different revision number (`1.13.5-meiss.4`) so that you know you've installed it.

Either:

-  `swiftlint` isn't getting called at all for you
- `swiftlint` is getting called but not returning anything
- `swiftlint` is getting called with the default or some other config that doesn't find any violations in the code.

**Comment by bealex** [comment]
@Javarome I've found a strange thing, when SwiftLint was not called completely if the file has other errors. Please try 1.13.6 (is being approved by JB now), maybe this is your case?
